File: ios/Playback/PreviewPlayer.swift
import AVFoundation
import Foundation

final class PreviewPlayer {
  static let shared = PreviewPlayer()

  private var player: AVAudioPlayer?
  private var activeSoundId: String?
  private let queue = DispatchQueue(label: "com.noirly.ringtonekit.preview")

  var onStateChanged: (([String: Any]) -> Void)?

  func play(soundId: String, uriString: String) throws {
    try queue.sync {
      stopInternal(emitStopped: true)

      guard let url = URL(string: uriString) else {
        throw RingtoneKitError(.invalidArgument, "Invalid sound URI: \(uriString)")
      }

      let resolvedURL: URL
      if url.isFileURL {
        resolvedURL = url
      } else if FileManager.default.fileExists(atPath: uriString) {
        resolvedURL = URL(fileURLWithPath: uriString)
      } else {
        throw RingtoneKitError(.soundNotFound, "Sound file not found: \(uriString)")
      }

      do {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playback, mode: .default, options: [])
        try session.setActive(true)
      } catch {
        throw RingtoneKitError(
          .playbackFailed,
          "Failed to activate audio session",
          nativeMessage: error.localizedDescription
        )
      }

      do {
        let audioPlayer = try AVAudioPlayer(contentsOf: resolvedURL)
        audioPlayer.delegate = PreviewPlayerDelegate.shared
        PreviewPlayerDelegate.shared.owner = self
        audioPlayer.prepareToPlay()
        activeSoundId = soundId
        player = audioPlayer
        guard audioPlayer.play() else {
          throw RingtoneKitError(.playbackFailed, "AVAudioPlayer failed to start")
        }
        emit(soundId: soundId, status: "started")
      } catch let error as RingtoneKitError {
        activeSoundId = nil
        player = nil
        emit(soundId: soundId, status: "error", errorMessage: error.message)
        throw error
      } catch {
        activeSoundId = nil
        player = nil
        emit(soundId: soundId, status: "error", errorMessage: error.localizedDescription)
        throw RingtoneKitError(
          .playbackFailed,
          "Failed to preview sound",
          nativeMessage: error.localizedDescription
        )
      }
    }
  }

  func stop() {
    queue.sync {
      stopInternal(emitStopped: true)
    }
  }

  fileprivate func handleFinished() {
    queue.async {
      if let soundId = self.activeSoundId {
        self.emit(soundId: soundId, status: "completed")
      }
      self.stopInternal(emitStopped: false)
    }
  }

  private func stopInternal(emitStopped: Bool) {
    let soundId = activeSoundId
    player?.stop()
    player = nil
    if emitStopped, let soundId {
      emit(soundId: soundId, status: "stopped")
    }
    activeSoundId = nil
    try? AVAudioSession.sharedInstance().setActive(false, options: [.notifyOthersOnDeactivation])
  }

  private func emit(soundId: String, status: String, errorMessage: String? = nil) {
    var payload: [String: Any] = [
      "soundId": soundId,
      "status": status,
    ]
    payload["errorMessage"] = errorMessage ?? NSNull()
    onStateChanged?(payload)
  }
}

private final class PreviewPlayerDelegate: NSObject, AVAudioPlayerDelegate {
  static let shared = PreviewPlayerDelegate()
  weak var owner: PreviewPlayer?

  func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
    owner?.handleFinished()
  }
}

File: android/src/main/java/com/noirly/ringtonekit/playback/PreviewPlayer.kt
package com.noirly.ringtonekit.playback

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.Ringtone
import android.media.RingtoneManager
import android.net.Uri
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.noirly.ringtonekit.errors.RingtoneKitException

class PreviewPlayer(
  private val context: Context,
  private val emit: (WritableMap) -> Unit,
) {
  private var ringtone: Ringtone? = null
  private var mediaPlayer: MediaPlayer? = null
  private var activeSoundId: String? = null

  @Synchronized
  fun play(soundId: String, uriString: String) {
    stopInternal(emitStopped = true)

    val uri = try {
      Uri.parse(uriString)
    } catch (error: Throwable) {
      throw RingtoneKitException.InvalidArgument("Invalid sound URI: $uriString")
    }

    activeSoundId = soundId

    try {
      if (uriString.startsWith("content://")) {
        val tone = RingtoneManager.getRingtone(context, uri)
          ?: throw RingtoneKitException.SoundNotFound("Ringtone not found for $uriString")
        tone.audioAttributes = AudioAttributes.Builder()
          .setUsage(AudioAttributes.USAGE_ALARM)
          .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
          .build()
        ringtone = tone
        tone.play()
        emitState(soundId, "started")
      } else {
        val player = MediaPlayer().apply {
          setAudioAttributes(
            AudioAttributes.Builder()
              .setUsage(AudioAttributes.USAGE_MEDIA)
              .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
              .build(),
          )
          setDataSource(context, uri)
          setOnCompletionListener {
            emitState(soundId, "completed")
            stopInternal(emitStopped = false)
          }
          setOnErrorListener { _, what, extra ->
            emitState(soundId, "error", "MediaPlayer error what=$what extra=$extra")
            stopInternal(emitStopped = false)
            true
          }
          prepare()
          start()
        }
        mediaPlayer = player
        emitState(soundId, "started")
      }
    } catch (error: RingtoneKitException) {
      activeSoundId = null
      emitState(soundId, "error", error.message)
      throw error
    } catch (error: Throwable) {
      activeSoundId = null
      emitState(soundId, "error", error.message)
      throw RingtoneKitException.PlaybackFailed(
        "Failed to preview sound",
        nativeMessage = error.message,
      )
    }
  }

  @Synchronized
  fun stop() {
    stopInternal(emitStopped = true)
  }

  @Synchronized
  private fun stopInternal(emitStopped: Boolean) {
    val soundId = activeSoundId
    try {
      ringtone?.stop()
    } catch (_: Throwable) {
    }
    ringtone = null

    try {
      mediaPlayer?.setOnCompletionListener(null)
      mediaPlayer?.setOnErrorListener(null)
      if (mediaPlayer?.isPlaying == true) {
        mediaPlayer?.stop()
      }
      mediaPlayer?.release()
    } catch (_: Throwable) {
    }
    mediaPlayer = null

    if (emitStopped && soundId != null) {
      emitState(soundId, "stopped")
    }
    activeSoundId = null
  }

  private fun emitState(soundId: String, status: String, errorMessage: String? = null) {
    val map = Arguments.createMap().apply {
      putString("soundId", soundId)
      putString("status", status)
      if (errorMessage != null) {
        putString("errorMessage", errorMessage)
      } else {
        putNull("errorMessage")
      }
    }
    emit(map)
  }
}
